### PR TITLE
FED-1757 Analyzer plugin null safety

### PR DESCRIPTION
**Pull request by sydneyjodon-wk**
## Motivation
  

We need to update the analyzer plugin to null safety

## Changes
  
- Copy tests to run on null safe and non-null safe code
- Copy playground examples to manually QA null safe vs non-null safe code
- Make minor updates for null safety:
  - `useEffect` dependency lists should allow nullable values
  - invalid child lint needs to work with nullable types
  - add ref assists should add nullable types where not dynamic

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Frontend Frameworks Design member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: tools/analyzer_plugin/pubspec.yaml [hunk 1]
```diff
@@ -53,3 +53,8 @@ workiva:
 # dependency_overrides:
 #   over_react:
 #     path: /Users/me/workspaces/over_react
+dependency_overrides:
+  over_react:
+    git:
+      url: https://github.com/Workiva/over_react.git
+      ref: analyzer_plugin_null_safety
```
Comment (sydneyjodon-wk) [inline tools/analyzer_plugin/pubspec.yaml:60]: todo I think I need to remove these before this merges?

File: tools/analyzer_plugin/test/integration/diagnostics/callback_ref_test.dart [hunk 2]
```diff
@@ -244,6 +353,7 @@ class CallbackRefDiagnosticFnComponentTestNoFix extends CallbackRefDiagnosticTes
 @reflectiveTest
 class CallbackRefDiagnosticClassComponentTest extends CallbackRefDiagnosticWithFixTest {
   static const usageSourceWithinClassComponent = '''
+// @dart=2.11
```
Comment (aaronlademann-wf) [inline tools/analyzer_plugin/test/integration/diagnostics/callback_ref_test.dart:356]: Is this for a non-null safe assertion / test somewhere?

File: tools/analyzer_plugin/test/integration/diagnostics/exhaustive_deps_test_cases.dart [hunk 3]
```diff
@@ -654,8 +654,8 @@ final Map<String, List<Map<String, Object>>> tests = {
           final maybeRef1 = useSomeOtherRefyThing();
           var state1 = useState<dynamic>(null);
           var state2 = over_react.useState<dynamic>(null);
-          var state3 = useReducer(null, null);
-          var state4 = over_react.useReducer(null, null);
+          var state3 = useReducer((_, __) => null, null);
```
Comment (aaronlademann-wf) [inline tools/analyzer_plugin/test/integration/diagnostics/exhaustive_deps_test_cases.dart:657]: Is this related to the null safe stuff? Or just a typo from the original implementation?

**Review by aaronlademann-wf — APPROVED** [review]
+1 with a few questions

**Review by greglittlefield-wf — CHANGES_REQUESTED** [review]
Whoops I forgot that I reviewed a good chunk of this but never finished; posting the comments I had so far, and will try to make a complete pass on Monday, sorry about that!

File: lib/src/component/hooks.dart [hunk 4]
```diff
@@ -112,7 +112,7 @@ StateHook<T> useStateLazy<T>(T Function() init) => react_hooks.useStateLazy<T>(i
 /// ```
 ///
 /// See: <https://reactjs.org/docs/hooks-effect.html#tip-optimizing-performance-by-skipping-effects>.
-void useEffect(dynamic Function() sideEffect, [List<Object>? dependencies]) => react_hooks.useEffect(sideEffect, dependencies);
+void useEffect(dynamic Function() sideEffect, [List<Object?>? dependencies]) => react_hooks.useEffect(sideEffect, dependencies);
```
[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply lib/src/component/hooks.dart:115]: Yup, this impl had a typo (it doesn't match the react-dart typings) and IIRC it was causing analysis issues in some of the test cases. Dependencies can totally be nullable.

File: tools/analyzer_plugin/lib/src/diagnostic/invalid_child.dart [hunk 5]
```diff
@@ -69,12 +71,13 @@ class InvalidChildDiagnostic extends ComponentUsageDiagnosticContributor {
       await validateReactChildType(argument.staticType, result.typeSystem, result.typeProvider,
           onInvalidType: (invalidType) async {
         final location = result.locationFor(argument);
+        final withNullability = result.unit.declaredElement?.library.featureSet.isEnabled(Feature.non_nullable) ?? false;
```
[the remark this answers is not in the document]
Comment (sydneyjodon-wk) [reply tools/analyzer_plugin/lib/src/diagnostic/invalid_child.dart:74]: Good idea! I like that!

File: tools/analyzer_plugin/lib/src/diagnostic/invalid_child.dart [hunk 6]
```diff
@@ -94,28 +97,26 @@ Future<void> validateReactChildType(DartType? type, TypeSystem typeSystem, TypeP
...
-  if (type.isDartCoreString) return;
+  if (nonNullType.isDynamic || nonNullType.isDartCoreObject) return;
+  if (nonNullType.isReactElement) return;
+  if (nonNullType.isDartCoreString) return;
   // isSubtypeOf to handle num, int, and double
-  if (typeSystem.isSubtypeOf(type, typeProvider.numType)) return;
-  if (type.isDartCoreNull) return;
-  if (type.isDartCoreBool) return;
+  if (typeSystem.isSubtypeOf(nonNullType, typeProvider.numType)) return;
+  if (nonNullType.isDartCoreNull) return;
```
[the remark this answers is not in the document]
Comment (sydneyjodon-wk) [reply tools/analyzer_plugin/lib/src/diagnostic/invalid_child.dart:108]: lol

File: tools/analyzer_plugin/test/integration/diagnostics/callback_ref_test.dart (hunk 2 above)
Comment (sydneyjodon-wk) [reply tools/analyzer_plugin/test/integration/diagnostics/callback_ref_test.dart:356]: Oh oops thanks for catching that!

**Review by greglittlefield-wf — APPROVED** [review]
- Changes look good
- Non-null-safe test copies are in-sync with null-safe ones
- Things work in null-safe and non-null-safe files in playground
- New tests run in CI, all tests are passing

+10

**Pull request merged**